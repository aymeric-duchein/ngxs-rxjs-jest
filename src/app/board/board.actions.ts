import {Issue, IssueStatus} from '../core/issue/issue.model';

export class BoardIssueClick {
  static readonly type = '[Board] Issue click';

  constructor(public issue: Issue) {
  }
}

export class BoardIssueDropped {
  static readonly type = '[Board] Issue dropped';

  constructor(public droppedIssue: Issue, public status: IssueStatus) {
  }
}
