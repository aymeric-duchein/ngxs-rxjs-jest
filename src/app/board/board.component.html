<div cdkDropListGroup class="board-list-container">
  <app-list class="mat-elevation-z2" [issues]="todoIssues$ | async" [status]="IssueStatus.TODO"></app-list>
  <app-list class="mat-elevation-z2" [issues]="inPRIssues$ | async" [status]="IssueStatus.IN_PROGRESS"></app-list>
  <app-list class="mat-elevation-z2" [issues]="doneIssues$ | async" [status]="IssueStatus.DONE"></app-list>
</div>

<app-add></app-add>

