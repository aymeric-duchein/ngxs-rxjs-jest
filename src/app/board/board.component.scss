@import '~@angular/material/theming';

@include mat-core();

app-add {
  position: absolute;
  bottom: 1rem;
  right: 3rem;
}

.board-list-container {
  height: calc(100vh - 104px);
  padding: 20px 10px;
  color: rgba(0, 0, 0, 0.87);
  display: flex;
  flex-direction: row;
  align-items: stretch;
  justify-content: space-between;
  font-size: 14px;
  position: relative;

  app-list {
    min-width: 30vw;
    max-width: 30vw;
    height: inherit;
    background: #fafafa;
    border-radius: 10px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }
}

