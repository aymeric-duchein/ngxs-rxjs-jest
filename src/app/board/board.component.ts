import {Component} from '@angular/core';
import {Issue, IssueStatus} from '../core/issue/issue.model';
import {IssueState} from '../core/issue/issue.state';
import {Observable} from 'rxjs';
import {Select} from '@ngxs/store';

export const doneIssueSelector = IssueState.getIssuesByStatus(IssueStatus.DONE);
export const todoIssueSelector = IssueState.getIssuesByStatus(IssueStatus.TODO);
export const inProgressIssueSelector = IssueState.getIssuesByStatus(IssueStatus.IN_PROGRESS);

@Component({
  selector: 'app-board',
  templateUrl: './board.component.html',
  styleUrls: ['./board.component.scss']
})
export class BoardComponent {
  private readonly IssueStatus = IssueStatus;

  @Select(doneIssueSelector) doneIssues$: Observable<Issue[]>;
  @Select(todoIssueSelector) todoIssues$: Observable<Issue[]>;
  @Select(inProgressIssueSelector) inPRIssues$: Observable<Issue[]>;

  constructor() {
  }
}

