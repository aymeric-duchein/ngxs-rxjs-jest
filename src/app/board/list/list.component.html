<h3>{{ status }}</h3>
<div
  class="list"
  cdkDropList
  [cdkDropListData]="issues"
  (cdkDropListDropped)="drop($event)">
  <app-issue cdkDrag *ngFor="let issue of issues" [issue]="issue" [cdkDragData]="issue"
             (issueClick)="issueClick($event)"></app-issue>
</div>
