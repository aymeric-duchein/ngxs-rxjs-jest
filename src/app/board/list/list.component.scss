div {
  cursor: pointer;
  flex: auto;
  overflow: auto;

  &::-webkit-scrollbar {
    width: 5px;
    height: 8px;
    background: rgba(9, 30, 66, .08);
    border-radius: 3px;
  }

  &::-webkit-scrollbar-thumb {
    background-color: #bfc4ce;
    border-radius: 3px;
  }
}

h3 {
  margin-left: 1em;
}

:host {
  padding-bottom: 10px;
  padding-left: 3px;
  padding-right: 3px;
}

app-issue {
  cursor: pointer;
}

