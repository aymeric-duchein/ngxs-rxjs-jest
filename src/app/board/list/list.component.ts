import {Component, Input} from '@angular/core';
import {Issue, IssueStatus} from '../../core/issue/issue.model';
import {CdkDragDrop} from '@angular/cdk/drag-drop';
import {Store} from '@ngxs/store';
import {BoardIssueClick, BoardIssueDropped} from '../board.actions';

@Component({
  selector: 'app-list',
  templateUrl: './list.component.html',
  styleUrls: ['./list.component.scss']
})
export class ListComponent {

  @Input() status: IssueStatus;
  @Input() issues: Issue[];

  constructor(private readonly store: Store) {
  }

  drop(event: CdkDragDrop<Issue[], Issue[]>) {
    this.store.dispatch(new BoardIssueDropped(event.item.data, this.status));
  }

  issueClick(issue: Issue) {
    this.store.dispatch(new BoardIssueClick(issue));
  }
}
