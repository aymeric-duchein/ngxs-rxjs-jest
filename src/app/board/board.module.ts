import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {BoardComponent} from './board.component';
import {ListComponent} from './list/list.component';
import {DragDropModule} from '@angular/cdk/drag-drop';
import {IssueModule} from '../shared/issue/issue.module';
import {AddModule} from '../shared/add/add.module';

@NgModule({
  declarations: [BoardComponent, ListComponent],
  imports: [
    CommonModule,
    DragDropModule,
    IssueModule,
    AddModule,
  ]
})
export class BoardModule {
}
