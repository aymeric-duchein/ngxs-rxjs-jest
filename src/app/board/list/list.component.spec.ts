import {ComponentFixture, TestBed} from '@angular/core/testing';
import {ListComponent} from './list.component';
import {NgxsModule, Store} from '@ngxs/store';
import {MockComponent, MockHelper, MockModule, MockRender} from 'ng-mocks';
import {CdkDrag, CdkDropList, DragDropModule} from '@angular/cdk/drag-drop';
import {IssueComponent} from '../../shared/issue/issue.component';
import {Issue, IssueStatus} from '../../core/issue/issue.model';
import {By} from '@angular/platform-browser';
import {DebugElement} from '@angular/core';
import {BoardIssueClick, BoardIssueDropped} from '@src/app/board/board.actions';
import {MOCK_ISSUES} from '@mocks/issue.mock';

describe('ListComponent', () => {
  let component: { issues: Issue[], status: IssueStatus };
  let fixture: ComponentFixture<{ issues: Issue[], status: IssueStatus }>;
  let store: Store;
  const mockIssues: Issue[] = MOCK_ISSUES;
  const mockStatus: IssueStatus = IssueStatus.TODO;
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        ListComponent,
        MockComponent(IssueComponent)
      ],
      imports: [
        NgxsModule.forRoot(),
        MockModule(DragDropModule)
      ]
    });

    fixture = MockRender('<app-list [issues]="issues" [status]="status"></app-list>',
      {
        issues: mockIssues,
        status: mockStatus
      });
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    jest.spyOn(store, 'dispatch');

    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  test('should display status', () => {
    const status: DebugElement = fixture.debugElement.query(By.css('h3'));
    expect(status).not.toBeNull();
    expect(status.nativeElement.innerHTML).toContain(IssueStatus.TODO);
  });

  test('should display a list of issues', () => {
    const issues = fixture.debugElement.queryAll(By.css('app-issue'));
    expect(issues.length).toBe(MOCK_ISSUES.length);
  });

  test('should initiate cdkDropList', () => {
    const dropList = fixture.debugElement.query(By.css('div[cdkDropList]'));
    expect(dropList).not.toBeNull();
    const cdkDropListDirective = MockHelper.getDirective(dropList, CdkDropList);
    expect(cdkDropListDirective.data).toEqual(MOCK_ISSUES);
  });

  test('should dispatch action on drop event', () => {
    const dropList = fixture.debugElement.query(By.css('div[cdkDropList]'));
    dropList.triggerEventHandler('cdkDropListDropped', {item: {data: MOCK_ISSUES[2]}});
    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(new BoardIssueDropped(MOCK_ISSUES[2], mockStatus));
  });

  test('should allow drag of issue', () => {
    const issue = fixture.debugElement.query(By.css('app-issue'));
    const cdkDragDirective = MockHelper.getDirective(issue, CdkDrag);
    expect(cdkDragDirective).not.toBeNull();
  });

  test('should pass correct data to drag', () => {
    const issue = fixture.debugElement.query(By.css('app-issue'));
    const cdkDragDirective = MockHelper.getDirective(issue, CdkDrag);
    const issueComponent = issue.componentInstance as IssueComponent;
    expect(issueComponent.issue).toEqual(MOCK_ISSUES[0]);
    expect(cdkDragDirective.data).toEqual(issueComponent.issue);
  });

  test('should dispatch action on issue click', () => {
    const issue = fixture.debugElement.query(By.css('app-issue'));
    issue.triggerEventHandler('issueClick', MOCK_ISSUES[2]);
    fixture.detectChanges();

    expect(store.dispatch).toHaveBeenCalledWith(new BoardIssueClick(MOCK_ISSUES[2]));
  });
});
