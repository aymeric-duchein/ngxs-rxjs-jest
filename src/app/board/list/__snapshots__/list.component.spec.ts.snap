// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListComponent should match snapshot 1`] = `
<mock-render
  issues={[Function Array]}
  status={[Function String]}
>
  <app-list
    ng-reflect-issues="[object Object],[object Object"
    ng-reflect-status="To do"
  >
    <h3>
      To do
    </h3>
    <div
      cdkdroplist=""
      class="list"
      ng-reflect-data="[object Object],[object Object"
    >
      
      <app-issue
        cdkdrag=""
        ng-reflect-data="[object Object]"
        ng-reflect-issue="[object Object]"
      />
      
      <app-issue
        cdkdrag=""
        ng-reflect-data="[object Object]"
        ng-reflect-issue="[object Object]"
      />
      
      <app-issue
        cdkdrag=""
        ng-reflect-data="[object Object]"
        ng-reflect-issue="[object Object]"
      />
      
    </div>
  </app-list>
</mock-render>
`;
